### PR TITLE
add camera orientation widget

### PR DESCRIPTION
**Pull request by akaszynski**
Add the camera orientation widget [vtk.CameraOrientationWidget](https://vtk.org/doc/nightly/html/classvtkCameraOrientationWidget.html)

This feature requires at least VTK 9.1.0

tmp

### CODE REVIEW

File: tests/test_widgets.py [hunk 1]
```diff
@@ -5,11 +5,17 @@
...
-NO_PLOTTING = not system_supports_plotting()
 mesh = examples.load_uniform()
 
+# skip all tests if unable to render
+if not system_supports_plotting():
+    pytestmark = pytest.mark.skip
+
+AT_LEAST_VTK_9_1_0 = False
+if pyvista._vtk.VTK9:
+    AT_LEAST_VTK_9_1_0 = pyvista._vtk.vtkVersion().GetVTKMinorVersion() >= 1
```
Comment (MatthewFlamm) [inline tests/test_widgets.py:16]: I think we should generalize this too. This pattern will become unmaintainable.  That could be a separate PR, especially given that this is in a test.

File: pyvista/plotting/widgets.py [hunk 2]
```diff
@@ -1730,6 +1732,49 @@ def _the_callback(widget, event):
...
+        >>> import pyvista
+        >>> mesh = pyvista.Cube()
+        >>> plotter = pyvista.Plotter()
+        >>> plotter.add_mesh(mesh, scalars=range(6), show_scalar_bar=False)
+        >>> plotter.add_camera_orientation_widget()
+        >>> plotter.show()
+
+        """
+        try:
+            from vtkmodules.vtkInteractionWidgets import vtkCameraOrientationWidget
```
Comment (MatthewFlamm) [inline pyvista/plotting/widgets.py:1761]: This no longer follows the usage of `_vtk` is there any way to generalize this as I also have filters that I'd like to add from 9.1.0?  Unless this is generalized, all new features will be ad hoc like this.

**Comment by MatthewFlamm** [comment]
I like this widget so much I would want it as a default in all my plotters, is there any way to set that up in a theme?  Sorry for front running a discussion feature request in the PR.

File: pyvista/plotting/widgets.py (hunk 2 above)
Comment (akaszynski) [reply pyvista/plotting/widgets.py:1761]: There's (at least) three ways of doing this:

1. Leave as is. It's adhoc, but means we don't need to keep our functionality in `_vtk`. Probably not the best approach

2. We could create an empty placeholder for VTK 9.1.0 compatibility as in
https://github.com/pyvista/pyvista/blob/363bca639c9050b2be5832e5105687691c43ea45/pyvista/_vtk.py#L481-L487
and then import always.

3.  Attempt to import and pass.

I think (2) will probably be the most robust.

---
Side note: maintaining backwards compatibility with vtk within pyvista has always been a goal of the project, since requiring a newer version of VTK has always been a non-goal since we want pyvista to be widely available. This is hard to do and we should try to do it right.

File: tests/test_widgets.py (hunk 1 above)
Comment (akaszynski) [reply tests/test_widgets.py:16]: What about (finally) adding in a vtk version?  For example:
```py
>>> import pyvista
>>> pyvista.vtk_version
(9, 1, 0)
```

**Comment by akaszynski** [comment]
> I like this widget so much I would want it as a default in all my plotters, is there any way to set that up in a theme? Sorry for front running a discussion feature request in the PR.

There is. We could simply enable this as an option within the `global_theme`:
```
>>> import pyvista
>>> pyvista.global_theme.enable_camera_orientation_widget = True
>>> pyvista.global_theme.enable_camera_orientation_widget
True

```

File: pyvista/_vtk.py [hunk 3]
```diff
@@ -451,3 +451,25 @@ def __init__(self):  # pragma: no cover
...
+# Get the VTK version. The API here is consistent since at least VTK v4.0.0
+try:
+    _ver = vtkVersion()
+    vtk_version = (
+        _ver.GetVTKMajorVersion(),
+        _ver.GetVTKMinorVersion(),
+        _ver.GetVTKBuildVersion()
+    )
+except:
+    (4, 0, 0)
```
Comment (MatthewFlamm) [inline pyvista/_vtk.py:475]: My understanding is that in `pyvista.__init__` that it will raise an error if the version is less than 5.x.x.  If we are sure that this works for all versions 4.0.0+, then why do we need this try...except?

It is also probably fine to leave this as is in case there are some edge cases so that users can still proceed.  But I would add a warning here and maybe set the version to be `(0, 0, 0)` to make it really obvious.

File: tests/test_widgets.py (hunk 1 above)
Comment (adeak) [reply tests/test_widgets.py:16]: With dev wheels and some peculiar vtk version choices (paraview??), I think we should go with something closer to an stdlib `version_info`. Namedtuple? A more complex object? Not sure how a dev wheel would fit in this in a way that still makes
```
if pyvista.vtk_version >= (9, 0)
```
work in a sane way.

File: pyvista/plotting/widgets.py (hunk 2 above)
Comment (adeak) [reply pyvista/plotting/widgets.py]: In that case we could have a `_vtk_compat.py` or similar, where all these placeholders are. In `_vtk` we can do
```py
try:
    from vtk.vtkSuperSpecificSubmodule import vtkUsefulFunctionality
except ImportError:
    from ._vtk_compat import vtkUsefulFunctionality
```

Any obvious problem with this approach?

File: pyvista/_vtk.py (hunk 3 above)
Comment (adeak) [reply pyvista/_vtk.py:475]: Sounds like thereás an assignment missing
```suggestion
    vtk_version = (4, 0, 0)
```

File: pyvista/plotting/widgets.py (hunk 2 above)
Comment (MatthewFlamm) [reply pyvista/plotting/widgets.py]: > Side note: maintaining backwards compatibility with vtk within pyvista has always been a goal of the project, since requiring a newer version of VTK has always been a non-goal since we want pyvista to be widely available. This is hard to do and we should try to do it right.

Right now, I think we are only testing with the vtk==8.2 and latest vtk.  We need to add back vtk==9.0.x.

I like @adeak suggestion, but I would generalize even further:

```py
from ._vtk_compat import vtkVersionError

try:
    from vtk.vtkSuperSpecificSubmodule import vtkUsefulFunctionality
except ImportError:
    vtkUsefulFunctionality = vtkVersionError("vtkUsefulFunctionality only supported for vtk versions >=9.0.0")
    
``` 

Then we only need 1 placeholder for all functions (or maybe some small number for different kinds).

Comment (adeak) [reply pyvista/plotting/widgets.py]: >  We need to add back vtk==9.0.x.
+1

>  I would generalize even further:

Good point. If we end up having too many of these we'll have to resist the temptation to do this dynamically in a loop :P

File: tests/test_widgets.py (hunk 1 above)
Comment (akaszynski) [reply tests/test_widgets.py]: Implemented in 9b631903af5ab40cfaf3113de90b17339f5ba832 as `pyvista.vtk_version_info`.

As far as I can tell, dev wheels will still work with:
```py
pyvista.vtk_version_info >= (9, 0)
```
Provided the base VTK release is 9.0.X****

File: pyvista/plotting/widgets.py (hunk 2 above)
Comment (akaszynski) [reply pyvista/plotting/widgets.py]: Can we implement this in a follow up PR? I wanted to limit the scope of the PR just to adding the widget, but it seems like whenever I implement a feature there are about 10 issues I uncover (probably my own doing...).

Comment (MatthewFlamm) [reply pyvista/plotting/widgets.py]: > 

I'm fine with this approach.  I wanted to highlight it since this is the first PR with features from 9.1.0.

File: pyvista/utilities/misc.py [hunk 4]
```diff
@@ -17,3 +18,123 @@ class PyvistaDeprecationWarning(Warning):
...
+
+    def __eq__(self, other):
+        """Return tuple-like equal to."""
+        return self._as_tuple() == other
+
+    def __iter__(self):
+        """Yield tuple-like iterator."""
+        yield self._as_tuple()
+
+    def __contains__(self, other):
```
Comment (MatthewFlamm) [inline pyvista/utilities/misc.py:131]: I'm not sure what the use case of this is.  Based on the test, we can do `assert 9 in pyvista.vtk_version`.  This seems not useful to me as it checks whether any value is the same.

File: pyvista/utilities/misc.py [hunk 5]
```diff
@@ -17,3 +18,123 @@ class PyvistaDeprecationWarning(Warning):
...
+    def __init__(self):
+        """Initialize the vtk version from the VTK class."""
+        # Get the VTK version. The API here is consistent since at least VTK v4.0.0
+        try:
+            ver = _vtk.vtkVersion()
+            self._major = ver.GetVTKMajorVersion()
+            self._minor = ver.GetVTKMinorVersion()
+            self._micro = ver.GetVTKBuildVersion()
+        except:  # pragma: no cover
+            self._major, self._minor, self._micro = (4, 0, 0)
```
Comment (MatthewFlamm) [inline pyvista/utilities/misc.py:62]: I think a warning here makes sense.  PyVista doesn't even support v4.0.0 of vtk, so it will be confusing if someone ever finds the version set to `(4, 0, 0)` without a warning.

**Review by MatthewFlamm — COMMENTED** [review]
I am good with the current PR state, a few last comments from me.

File: pyvista/utilities/misc.py (hunk 4 above)
Comment (akaszynski) [reply pyvista/utilities/misc.py:131]: I just wanted to make this like `sys.version_info`, where they're `tuple`-like. I agree that we don't need this, but this is mostly so we can make it look like the build-in python `version_info`.  I could go either way with this.

File: pyvista/utilities/misc.py (hunk 5 above)
Comment (akaszynski) [reply pyvista/utilities/misc.py:65]: Added a warning in 58e7aa18bb212fcbd21b9f225bf7c4cb86ae3a19.

File: tests/test_widgets.py (hunk 1 above)
Comment (adeak) [reply tests/test_widgets.py]: Yeah, I was worried about a potential string-valued micro version breaking comparison with a proper numerical micro version. It's likely that VTK's API only returns a numerical micro anyway. And we're unlikely to test for micro versions.

File: pyvista/utilities/misc.py (hunk 4 above)
Comment (MatthewFlamm) [reply pyvista/utilities/misc.py:134]: I'm fine if this mirrors the python implementation.

File: pyvista/utilities/misc.py [hunk 6]
```diff
@@ -17,3 +20,124 @@ class PyvistaDeprecationWarning(Warning):
     """Non-supressed Depreciation Warning."""
 
     pass
+
+
+class VTKVersionInfo():
```
Comment (adeak) [inline pyvista/utilities/misc.py:25]: Do we really need a full-fledged class for this? Almost all of this seems like a namedtuple, with a custom repr and the API check at the start. A factory that returns a namedtuple seems more concise, even if we have to monkeypatch its `__repr__` (but namedtuples are a hack anyway). Something like
```py
from collections import namedtuple

def VTKVersionInfo():
    version_info = namedtuple('VTKVersionInfo', ['major', 'minor', 'micro'])
    version_info.__repr__ = lambda vi: f"VTK Version {vi.major}.{vi.minor}.{vi.micro}"
    try:
        ver = _vtk.vtkVersion()
        major = ver.GetVTKMajorVersion()
        minor = ver.GetVTKMinorVersion()
        micro = ver.GetVTKBuildVersion()
    except:  # pragma: no cover
        warnings.warn("Unable to detect VTK version. Defaulting to v4.0.0")
        major, minor, micro = (4, 0, 0)
    return version_info(major, minor, micro)
```

This seems to work:
```py
>>> vi = VTKVersionInfo()
>>> vi
VTK Version 9.1.0
>>> vi.major
9
>>> vi.micro
0
>>> 9 in vi  # does this even make any sense?
True
```

File: pyvista/utilities/misc.py (hunk 4 above)
Comment (adeak) [reply pyvista/utilities/misc.py:134]: I don't think `__contains__` makes sense here, and the class is missing `__iter__` etc. so it's not a tuple anyway. I'd remove the needless bits. But I actually have a `namedtuple` recommendation in another comment.

File: pyvista/utilities/misc.py (hunk 6 above)
Comment (adeak) [reply pyvista/utilities/misc.py:25]: Oh, and I don't like that bare `except` at all. We should try being more specific; I suspect an `AttributeError` would be appropriate.

Comment (akaszynski) [reply pyvista/utilities/misc.py]: As I kept adding methods to make it tuple-like I was getting the feeling one of you would be critical about the implementation. I agree that this is quite a bit cleaner. I've removed the `__repr__` monkeypatch because it makes it look more like Python's `sys.version_info`.

**Review by MatthewFlamm — APPROVED** [review]
LGTM now

**Comment by emielke12** [comment]
image

Has anyone ever encountered this issue before?

**Comment by adeak** [comment]
@emielke12 I haven't used this widget yet, but do you have an example (which need not contain your real data) that produced that? You should probably open an issue about it and we can take a look.
